### ISSUE TITLE
stacks-node loses outbound peers when broadcasting events 

### ISSUE THREAD

**Issue by CharlieC3** [stacks-network/stacks-core#3145]
The stacks-node is susceptible to losing all outbound peers when broadcasting events to a stacks-blockchain-api, causing it to fall behind the chain tip potentially for hours until the stacks-node is restarted.
During this time, the stacks-node printing plenty of `PermanentlyDrained` logs like such:
`http:id=234,request=false,peer=<redacted>: failed to recv: PermanentlyDrained`

**Expected behavior**
If a stacks-node loses all outbound peers, it should be able to acquire new ones or re-validate old ones relatively quickly.

**Debug Information**
If this happens again, I'll get more debug information like:
* running `'select * from frontier;'` in the `peers.db` before/after restarting the stacks-node.
* checking open network connections before/after restarting the stacks-node

stacks-node version: `2.05.0.2.0`

<img width="1754" alt="Screen_Shot_2022-05-21_at_12 25 11_PM" src="https://user-images.githubusercontent.com/2747302/169853120-3425dd8b-c686-4b2e-adb2-9984aaef898e.png">

**Comment by wileyj** [comment]
one more thing to check is the DB load/queries when this happens

**Comment by jcnelson** [comment]
I'd also look at things originating from `src/net/neighbors.rs` on the `INFO`, `WARN`, or `ERRO` level.

**Comment by CharlieC3** [comment]
@wileyj According to our DB metrics, it was under normal load during the day which the stacks-node lost all outbound peers.

@jcnelson No `ERRO` level logs at all. However, I do see some of these logs around the time all outbound peers were lost (~1433 UTC):
`local.1://(bind=0.0.0.0:20444)(pub=None): Failed to begin neighbor walk from pingback: NoSuchNeighbor`

Here's some logs filtered to `src/net/neighbors.rs` between 1400 - 1500 UTC. The issue occurs around 1433 UTC:
[src-net-neighbors-3145.csv](https://github.com/stacks-network/stacks-blockchain/files/8756179/src-net-neighbors-3145.csv)

**Comment by wileyj** [comment]
> @wileyj According to our DB metrics, it was under normal load during the day which the stacks-node lost all outbound peers.
> 
> @jcnelson No `ERRO` level logs at all. However, I do see some of these logs around the time all outbound peers were lost (~1433 UTC): `local.1://(bind=0.0.0.0:20444)(pub=None): Failed to begin neighbor walk from pingback: NoSuchNeighbor`
> 
> Here's some logs filtered to `src/net/neighbors.rs` between 1400 - 1500 UTC. The issue occurs around 1433 UTC: [src-net-neighbors-3145.csv](https://github.com/stacks-network/stacks-blockchain/files/8756179/src-net-neighbors-3145.csv)

one more thing you could look at if you have the data - slow query log, to see if there were any slow queries around the time you noticed the issue. doesn't seem like the API DB is the problem though

**Comment by CharlieC3** [comment]
All I can find is a small blip where there was a slow query which took about 9 seconds to return around 1429 UTC on May 21st, but infrequent slow queries like those are not unheard of.
Depending on the query complexity, queries over 5 seconds could occur dozens of times a day.

Cross-comparing it to today's metrics, these queries don't appear to have any effect on the outbound peers for the live stacks-node broadcasting events.

**Comment by CharlieC3** [comment]
@deantchi and I found something pretty interesting. It seems the stacks-node we're running which broadcasts to the API wasn't the only node in our infrastructure to have lost all outbound peers.
![Screen Shot 2022-05-23 at 4 53 05 PM](https://user-images.githubusercontent.com/2747302/169905859-190824c9-bc41-4b90-b91c-f5f711ecbe0d.png)

From the screenshot, it's clear 3 other nodes lost all outbound peers. The difference is that the other 3 nodes were able to recover in about 5 minutes, while the stacks-node broadcasting events didn't recover for a few hours until it was restarted.

This screenshot is the same three nodes, but over a longer time period. You can see the stacks-node broadcasting events would get an outbound peer or two every so often, but would lose it pretty quickly.
This may explain why I noticed the API was still receiving mempool TXs periodically.
![Screen Shot 2022-05-23 at 4 55 18 PM](https://user-images.githubusercontent.com/2747302/169906248-57709998-77b2-4bbe-8fa5-511c7c7dd91b.png)

**Comment by jcnelson** [comment]
Thanks for the csv.  Can you b me one that also includes `src/net/chat.rs` and `src/net/p2p.rs`, in addition to `src/net/neighbors.rs`?

Also, to be sure -- you're using a different node private key on each node, right?  You can check via `/v2/info` and looking at `node_public_key` -- it should be different for each node.

**Comment by CharlieC3** [comment]
Here are the other log files. The `src/net/p2p.rs` log is quite long, so I had to trim the timeline down to 1420 - 1500 UTC.
[src-net-p2p-3145.csv](https://github.com/stacks-network/stacks-blockchain/files/8759077/src-net-p2p-3145.csv)
[src-net-chat-3145.csv](https://github.com/stacks-network/stacks-blockchain/files/8759078/src-net-chat-3145.csv)

> You're using a different node private key on each node, right?

These are just followers, not miners. Does the `seed` property need to be configured for each of them regardless?

**Comment by jcnelson** [comment]
I'm seeing lots of `HandshakeReject`s in `/src/net/chat.rs`, which indicates that your node is doing something wrong.  Unfortunately we can't see the reason since that's logged on the remote node doing the rejection.  But, the possible reasons are as follows:

* You're using the same `local_peer_seed` as the remote node, which will make the remote node think it's trying to handshake to itself (since it has the same signature as its own private key)

* The handshake's reported public key expiry block has passed (this is unlikely)

* Your node is an outbound peer of the remote node (i.e. the remote node reaches out to yours first), and the IP address your node reports is different than the one that the remote node used to contact you

**Comment by CharlieC3** [comment]
> You're using the same local_peer_seed as the remote node, which will make the remote node think it's trying to handshake to itself (since it has the same signature as its own private key)

If the `local_peer_seed` property is not defined in the config.toml, [doesn't the binary generate a random hex string to use at bootup](https://github.com/stacks-network/stacks-blockchain/blob/master/testnet/stacks-node/src/config.rs#L1306)? Unless I'm misunderstanding something, it should be near impossible for two nodes to generate the same `local_peer_seed` value, right?

> Your node is an outbound peer of the remote node (i.e. the remote node reaches out to yours first), and the IP address your node reports is different than the one that the remote node used to contact you

None of the nodes losing outbound peers are exposed publicly, so I don't see how they could be considered an outbound peer of an external node. They could likely only be outbound peers of other nodes within the same private network.

**Comment by CharlieC3** [comment]
Jude answered my question above in the weekly blockchain meeting. Posting answer for posterity.

> If the local_peer_seed property is not defined in the config.toml, [doesn't the binary generate a random hex string to use at bootup](https://github.com/stacks-network/stacks-blockchain/blob/master/testnet/stacks-node/src/config.rs#L1306-L1307)? Unless I'm misunderstanding something, it should be near impossible for two nodes to generate the same local_peer_seed value, right?

> The stacks-node will generate a `local_peer_seed` at bootup if started with an empty working directory. However since we boot our stacks-nodes with a snapshot, they all start with essentially the same working directory, and thus have the same `local_peer_seed` value, which may have been causing issues.

___

@jcnelson One of our stacks-nodes broadcasting events lost all outbound peers again last night for almost 4 hours until it was restarted. This stacks-node is running with a unique `local_peer_seed` configured, so it seems changing this property does not have a large enough effect to mitigate this problem.

Please let me know if there's any additional data I can help provide which would help in the investigation of this problem.

**Comment by jcnelson** [comment]
> This stacks-node is running with a unique local_peer_seed configured

Is this also the case in the peers DB?  i.e. the `local_peer` table reflects the unique peer seed?

**Comment by CharlieC3** [comment]
@jcnelson Hmm, unexpectedly, the entry in the `local_peer` table does _not_ appear to be unique despite `local_peer_seed` being set to a unique 32 character hex string. I'm guessing updating this entry could do the trick, but it doesn't seem to be doing so at bootup. Is this a bug?

**Comment by jcnelson** [comment]
Yeah, the node really should override the private key in `local_peer` with whatever is in `local_peer_seed` if explicitly set.  Wanna open a separate issue for that?

If you fix the `local_peer` table, does the problem go away?

**Comment by jcnelson** [comment]
The private key in `local_peer` should be a secp256k1 private key, with a `01` at the end to indicate that the public key should be compressed.  If `local_peer_seed` is neither 64 nor 66 bytes long, or is not a valid secp256k1 private key, then it will be hashed repeatedly until the resulting hash is a valid private key.

But either way, the problem I'm trying to rule out is that you have two or more peers with the same private key in their `local_peer` table.  If so, then that would explain this behavior.

**Comment by CharlieC3** [comment]
@jcnelson I see, thanks for the info. I replaced the db entry with a similar one, but with a different private key. I generated the  `secp256k1` key using the [python module](https://pypi.org/project/secp256k1/), and added `01` to the end of it like the original entry.

It seems to have been accepted. I've made this change to two stacks-nodes now and they're still progressing along the chain. I'll keep a close watch on these over the next couple of weeks for any loss of outbound peers.
